### PR TITLE
add STAT for ps.py

### PR DESCRIPTION
**Pull request by chenhanxiao**
Add STAT to display  the  state of a process.

Signed-off-by: Chen Hanxiao chenhanxiao@gmail.com

### CODE REVIEW

**Comment by giampaolo** [comment]
There's no need to modify _pslinux.py. The change should happen in ps.py only.

**Comment by chenhanxiao** [comment]
Ah, thanks.
I'll post a v2.

**Comment by chenhanxiao** [comment]
@giampaolo 
Any comments？
btw: this PR did not bring new ci failures.

**Comment by giampaolo** [comment]
I am flying to US. I will try to take a look at this sometime during next week.

**Comment by chenhanxiao** [comment]
@giampaolo 
ping

**Comment by chenhanxiao** [comment]
@giampaolo 
hi, any comments?

File: scripts/ps.py [hunk 1]
```diff
@@ -17,17 +17,43 @@
...
+STATUS_STOPPED = "stopped"
+STATUS_TRACING_STOP = "tracing-stop"
+STATUS_ZOMBIE = "zombie"
+STATUS_DEAD = "dead"
+STATUS_WAKE_KILL = "wake-kill"
+STATUS_WAKING = "waking"
+STATUS_IDLE = "idle"  # FreeBSD, OSX
+STATUS_LOCKED = "locked"  # FreeBSD
+STATUS_WAITING = "waiting"  # FreeBSD
+STATUS_SUSPENDED = "suspended"  # NetBSD
```
Comment (giampaolo) [inline scripts/ps.py]: there's no need to redefine all these; they are defined in the psutil namespaces aready (use `psutil.STATUS_SUSPENDED`, etc.)

File: scripts/ps.py [hunk 2]
```diff
@@ -17,17 +17,43 @@
...
+PROC_STATUSES_RAW = {
+    STATUS_RUNNING:             "R",
+    STATUS_SLEEPING:            "S",
+    STATUS_DISK_SLEEP:          "D",
+    STATUS_STOPPED:             "T",
+    STATUS_TRACING_STOP:        "t",
+    STATUS_ZOMBIE:              "Z",
+    STATUS_DEAD:                "X",
+    STATUS_WAKE_KILL:           "K",
+    STATUS_WAKING:              "W"
```
Comment (giampaolo) [inline scripts/ps.py]: these are not PEP-8 compliant. Redefine the above by removing unnecessary spaces (`STATUS_WAKING: "W"` etc.)

File: scripts/ps.py [hunk 3]
```diff
@@ -17,17 +17,28 @@
...
+PROC_STATUSES_RAW = {
+    psutil.STATUS_RUNNING: "R",
+    psutil.STATUS_SLEEPING: "S",
+    psutil.STATUS_DISK_SLEEP: "D",
+    psutil.STATUS_STOPPED: "T",
+    psutil.STATUS_TRACING_STOP: "t",
+    psutil.STATUS_ZOMBIE: "Z",
+    psutil.STATUS_DEAD: "X",
+    psutil.STATUS_WAKING: "W"
+}
```
Comment (giampaolo) [inline scripts/ps.py]: This list is incomplete. It's missing:

```
psutil.STATUS_WAKE_KILL
psutil.STATUS_WAKING
psutil.STATUS_IDLE(OSX, FreeBSD)
psutil.STATUS_LOCKED(FreeBSD)
psutil.STATUS_WAITING(FreeBSD)
psutil.STATUS_SUSPENDED(NetBSD)
```

You should also take that into account. You can probably come up with new letters.

File: scripts/ps.py [hunk 4]
```diff
@@ -64,6 +75,7 @@ def main():
                 int(pinfo['memory_info'].rss / 1024) or '?'
             memp = pinfo['memory_percent'] and \
                 round(pinfo['memory_percent'], 1) or '?'
+            stat = PROC_STATUSES_RAW.get(pinfo['status'], '')
```
Comment (giampaolo) [inline scripts/ps.py]: please rename the variable to "status". "stat" is a stdlib module and it's bad practice to obfuscate "reserved" names.

Comment (giampaolo) [inline scripts/ps.py]: Also if the letter is not available simply return the original `p['status']` as in:

```
PROC_STATUSES_RAW.get(pinfo['status'], pinfo['status'])
```

File: scripts/ps.py (hunk 3 above)
Comment (chenhanxiao) [inline scripts/ps.py]: I checked take man page of ps in linux for a reference.
Found nothing about the status above.
Could you give me some hints?

Comment (giampaolo) [inline scripts/ps.py]: These are from other UNIX platforms (FreeBSD, OSX). psutil exposes them also if you're on Linux. You should take all posible statuses into account.

Comment (chenhanxiao) [inline scripts/ps.py]: OK, I'll download them and have a look.
Thanks.

Comment (chenhanxiao) [inline scripts/ps.py]: I updated these:
psutil.STATUS_IDLE(OSX, FreeBSD):		I
psutil.STATUS_LOCKED(FreeBSD):		L
psutil.STATUS_WAITING(FreeBSD):			W
psutil.STATUS_SUSPENDED(NetBSD):			V

But:
psutil.STATUS_WAKE_KILL
psutil.STATUS_WAKING
the ps cmd of Linux did not take into account.

Maybe we should do the same thing.

Comment (giampaolo) [inline scripts/ps.py]: No we shouldn't. Come up with some letter. `psutil.STATUS_WAKE_KILL` can be mapped to `"WK" and `psutil.STATUS_WAKING` could be `"WA"`.

File: scripts/ps.py [hunk 5]
```diff
@@ -17,17 +17,33 @@
...
+    psutil.STATUS_STOPPED: "T",
+    psutil.STATUS_TRACING_STOP: "t",
+    psutil.STATUS_ZOMBIE: "Z",
+    psutil.STATUS_DEAD: "X",
+    psutil.STATUS_WAKING: "WA",
+    psutil.STATUS_IDLE: "I",
+    psutil._common.STATUS_WAKE_KILL: "WK",
+    psutil.STATUS_LOCKED: "L",
+    psutil.STATUS_WAITING: "W",
+    psutil._common.STATUS_SUSPENDED: "V"
```
Comment (giampaolo) [inline scripts/ps.py]: Don't do this. Instead do:

```
if hasattr(psutil, 'STATUS_SUSPENDED'):
    PROC_STATUSES_RAW[psutil.STATUS_SUSPENDED] = "V"
```

...same for `STATUS_WAKE_KILL`.

**Comment by giampaolo** [comment]
OK, looks good. Merging. Thanks.

**Pull request merged**